Varya Theme: Adding a css-variables version of the original version of Varia. (#3)
Automattic/themes-workspace: varya/README.md

# Varya

> Varya is a styling system for quickly creating Gutenberg-ready themes for WordPress.

Reviewer: Should this be rephrased to make clear that it's a parent theme itself?

Author: Yup, will do!

## What does it do?

> When you reduce a theme design down to a set of systematic design decisions, you end up with something called a _Style Guide_. The Varya system works by taking the rules of a Style Guide and expressing them through carefully placed variables or _design tokens_ that influence the appearance of a WordPress site. It currently unifyies the overall of Gutenberg Blocks, the theme Header + Footer areas, WooCommerce, Jetpack and more. It also syncs styles between the editor and the frontend so that you don’t need to hand-write styles for both. This greatly speeds up the Gutenberg theme development process and reduces the amount of manual styling that typically goes into developing a theme. 

Reviewer: > It currently unifyies the overall of Gutenberg Blocks, the theme Header + Footer areas, WooCommerce, Jetpack and more.

This doesn't totally make sense. I think it's missing a word or two?

### What controls does the system come with?

  - **Fonts** - Font-family, size, weight, and line-height rules. 
  - **Colors** - Primary, secondary, background, foreground and border colors. 

>   - **Spacing** - Sets an 8px vertical rhythm between all blocks and major components. Adds general spacing rules for blocks and components. Adds utility spacing classes for negative margins.

Reviewer: ```suggestion
  - **Spacing** - A default 8px vertical rhythm between all blocks and major components. It also includes utility spacing classes for negative margins.
```

>   - **Responsive Logic** - Unifies the responsive behavior across various Blocks and Components to simplify behaviors.

Reviewer: ```suggestion
  - **Responsive Logic** - Built-in responsive behavior across Blocks and Components.
```

## How does it work?

The system itself lives in the `/varya/sass` directory as a collection of Sass partials broken up by scope and hierarchy. The partials get compiled down to singular CSS files that live in the main Varya directory and cascade downward like so:

**Frontend** 
- `/varya/variables.css`
- `/varya/style.css`

**Editor** 
- `/varya/variables-editor.css`
- `/varya/style-editor.css`

**Customizer**

> - `/varya/variables.css`

Reviewer 2: Is there a reason the customizer variables and frontend variables are contained in the same partial?

Author: Woops! Yeah, that was left over from when Varya still had customizer color options. Now that those are gone, I can remove this section.

- `/varya/style.css`

In each view, the variables are loaded first and then the stylesheet is loaded which applies the variables.

> ## How to use it

Author: Fully agreed. 

I need to sync with @jeffikus to get the exact process as it relates to the build tool, but even with out the tool, the steps here should be more explicit than they are now. Good catch and will do.

To use the system, simply duplicate the `vayra-child` theme directory and rename it `my-theme-name`. You’ll also want to do a search for `varia-child` strings and replace them with `my-theme-name` as follows:

  - Search for: `'varya-child'` and replace with: `'my-theme-name'` (with quotes).
  - Search for: `varya_child_` and replace with: `my_theme_name_`.
  - Search for: `Text Domain: varia-child` and replace with: `Text Domain: my-theme-name` in _style.css_.
  - Search for:  `varia-child` and replace with: `my_theme_name`.
  - Search for: `varia-child-` and replace with: `my-theme-name-`.

Soon this process will be replaced by a `theme-dev-util` (similar to [this](https://github.com/Automattic/theme-dev-utils)) that allows you to run a command and automatically produce a child theme with all the strings already replaced.

## Simple Child Theme Structure 

> (See: `/varya-child`)

Reviewer 2: I may not understand fully, but isn't the child theme the actual theme you activate on your site? If so, should that be made more explicit?

Author: Yup, once I revise the “How To Use It” section this should become more clear.

Use this simple Varya child-theme example to familiarize yourself with the CSS-variables and to see how they influence the theme design when you change them. To start, open up the `/varya-child-simple/variables.css` file. Change any of the values, save the file, and refresh the frontend of your site to see the changes. To add more variables, view the [full list of available variables](https://github.com/Automattic/themes-workspace/blob/add/variatheme/varya/variables.css) from the Varya parent them and copy any variables you want in to your child theme. 
The stylesheet structure for a child-theme cascades downward like this:

### Frontend
- `/varya/variables.css`

> - `/varya-child-simple/variables.css` (System Override)

Reviewer: ```suggestion
- `/varya-child-simple/variables.css` (System Overrides)
```

- `/varya/style.css`
- `/varya-child-simple/style.css` (Extra CSS)

### Editor
- `/varya/variables-editor.css`
- `/varya-child-simple/variables-editor.css` (System Overrides)
- `/varya/style-editor.css`
- `/varya-child-simple/style-editor.css` (Extra CSS)

### System Overrides

A list of CSS-variables that override the variables in the child theme. This is where you tell the system to use the _Futura_ font-family instead of the _sans-serif_ default, for example. There’s no need to replace all of the variables here, only the ones you wish to actually change based on your theme design.

**Varya parent theme font style** (`/varya/variables.css`)

`:root { --global--font-family: sans-serif; }`

**Varya child-theme font-style overrides** (`/varya-child/variables.css`)

> `:root { --global--font-family: Futura, Helvetica; }`

Reviewer: ```suggestion
`:root { --global--font-family: "Futura", "Helvetica", sans-serif; }`
```

### Extra CSS

> These should be supplemental styles that give the theme a unique appearance beyond what’s possible with the Varya system. Need to add a fixed header or add a box-shadow to your theme’s buttons? This is where those styles would go. When possible and appropriate, try to include Varya CSS-variables in this styles sheet so that the system retains its usefulness across the theme.

Reviewer: ```suggestion
These should be supplemental styles that give the theme a unique appearance beyond what’s included with the Varya system defaults. Need to add a fixed header or add a box-shadow to your theme’s buttons? This is where those styles would go. When possible and appropriate, try to use Varya CSS-variables in this stylesheet so that the system retains its usefulness across the theme.
```

## Advanced Child Theme Structure 

> (See: `/varya-child-advanced`)

Reviewer 2: I feel like this should eventually be a separate example with its own repo?

Reviewer: Yeah, that's a good idea. But it's fine to keep this in here for now.

Author: Yeah, this is actually just the Varya version of “Nwaneri” so eventually it’ll be its own thing anyway.

Alternatively, I wonder if common code snippets might be useful to include here. With Varya themes you spend most of your time working on the header & footer UI since those areas are Gutenberg-able yet. 

If we were to come up with 2-3 common header patterns for example, we could create some Varya specific snippets that give you a jump start on styling those areas, at least until FSE is here.

| **Wide & up**     | `$breakpoint_xxl`   | 1280px       |                                                                                                                   |

Once you’ve set these variables to your liking in `/varya-child-advanced/responsive.scss`, run `npm run build` on the child-theme which will recompile `responsive.scss` and reset the content width CSS-variables in `responsive.css`.

## Todos
 - [DONE] Explore responsive-logic overrides in the child theme.
 - [DONE] Introduce an advanced child theme example that uses block styles
 - Audit Varya system for excessive overridden rules
 - Integrate with the child-theme build tool.
 - Add Jetpack support?
 - Optimize responsive font-size styles.
 - [DONE] Audit how variables appear in the editor, the customizer, and the frontend.
 - Consider adding defaults fallbacks to top-level css-variables.
 - Create a demo.
 - Add a CSS-Variables polyfill or other fallback solution for older browsers.
 - [WIP] Create a ReadMe that describes how to use it.

### License

> - GPL 2.0

Reviewer: This should link to the LICENSE file. Also, are there any other license disclosures that we should list here?

Reviewer: Actually... I think these are all missing license files. 😄

Author: Ahh, will do. 

May need to reference the image used in the screenshot here. Need to double check that it’s CC0, first but that would likely go here. I’ll compare with a few themes to see what else might need to be included here.
